### ISSUE TITLE
Nested Carthage Dependencies, not finding Framework Headers -- Framework Search Path?

### ISSUE THREAD

**Issue by fbartho** [Carthage/Carthage#1053]
When trying to use a framework through Carthage, that itself depends on another framework through Carthage, Xcode doesn't seem to find the nested framework's headers when building the leaf framework.

When I go into the Carthage/Checkouts directory, and open the leaf Xcodeproj, it too fails to find the built framework -- even though it builds the target dependency. Removing & Re-adding the dependent framework by manually dragging the built dependency in to the link stage does not fix the problem.

Workaround: manually adding `${PROJECT_DIR}/Carthage/Build/**` to the `Framework Search Paths` of the leaf dependency fixes the build.

Note: I'm new to Carthage, so please advise if I'm doing something wrong.
## Using
- Clean New Swift Project & Workspace
- Carthage 0.10.0 -- Also occurs with Carthage 0.11.0
- Xcode 7.1.1 -- Also occurs with Xcode 7.2
- OS X 10.11.2
## Transcript
1. Xcode GUI: create new iOS Tabbed Application (Path ~/Code/cookbook/cookbook). Change no settings at all.
2. Create .gitignore using standard settings + Carthage settings.
3. Create Cartfile
   `github "ParsePlatform/Parse-SDK-iOS-OSX" "master"`
4. In terminal:

```
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook$  ls -al
total 24
drwxr-xr-x   9 fbarthelemy  staff  306 Jan 18 12:19 .
drwxr-xr-x   5 fbarthelemy  staff  170 Jan 17 22:47 ..
drwxr-xr-x  13 fbarthelemy  staff  442 Jan 18 12:20 .git
-rw-r--r--@  1 fbarthelemy  staff  227 Jan 17 22:10 .gitignore
-rw-r--r--@  1 fbarthelemy  staff   50 Jan 17 22:33 Cartfile
drwxr-xr-x   8 fbarthelemy  staff  272 Jan 17 22:08 Cookbook
drwxr-xr-x   5 fbarthelemy  staff  170 Jan 17 22:08 Cookbook.xcodeproj
drwxr-xr-x@  6 fbarthelemy  staff  204 Jan 17 22:08 build
-rwxr-xr-x@  1 fbarthelemy  staff   68 Jan 17 23:46 cartupdate
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook$  carthage version
0.10.0
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook$  cat Cartfile
github "ParsePlatform/Parse-SDK-iOS-OSX" "master"
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook$  carthage update
*** Fetching Parse-SDK-iOS-OSX
*** Fetching Bolts-iOS
*** Checking out Parse-SDK-iOS-OSX at "f3522cfb5b72830810cbfc43458d2add0d363bcf"
*** Checking out Bolts-iOS at "1.6.0"
*** xcodebuild output can be found in /var/folders/vk/lwg7x4cn30ng2dw55_7536tr0000gn/T/carthage-xcodebuild.XKuxX2.log
*** Building scheme "Bolts-OSX" in Bolts.xcodeproj
*** Building scheme "Bolts-tvOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-watchOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-iOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Parse-watchOS-Dynamic" in Parse.xcworkspace
** BUILD FAILED **


The following build commands failed:
    CompileC build/Parse.build/Release-watchsimulator/Parse-watchOS-Dynamic.build/Objects-normal/i386/PFCommandURLRequestConstructor.o Parse/Internal/Commands/CommandRunner/URLRequestConstructor/PFCommandURLRequestConstructor.m normal i386 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
/Users/fbarthelemy/Code/cookbook/cookbook/Carthage/Checkouts/Parse-SDK-iOS-OSX/Parse/Internal/BFTask+Private.h:12:9: fatal error: 'Bolts/BFExecutor.h' file not found
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
    CompileC build/Parse.build/Release-watchsimulator/Parse-watchOS-Dynamic.build/Objects-normal/i386/PFCommandURLRequestConstructor.o Parse/Internal/Commands/CommandRunner/URLRequestConstructor/PFCommandURLRequestConstructor.m normal i386 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
```
### Potentially Related issues
- #79
- #773
- #68

**Comment by fbartho** [comment]
If it helps: I normally run with relative build folder at `${PROJECT_DIR}/build/` but switching back to default (DerivedData) did not resolve the problem.

**Comment by mdiep** [comment]
I'm not able to reproduce this with Carthage 0.11.

It may help to check for errors in the xcodebuild output.

**Comment by fbartho** [comment]
@mdiep Here is a link to an example xcodebuild output exhibiting this issue. http://cl.ly/3h1e2w3p1c1V

I can attempt this again with any combination of parameters, it has proven to happen persistently on my machine. Would you like me to run it with any given settings? What carthage commandline should I use? Is there a verbose option that I could provide to help debug this further?

[To recap: to repro this, I take an empty directory, I add `github "ParsePlatform/Parse-SDK-iOS-OSX"` as a Cartfile, finally, I run `carthage update`]

**Comment by mdiep** [comment]
Try deleting `~/Library/Caches/org.carthage.CarthageKit` and running it again.

> [To recap: to repro this, I take an empty directory, I add github "ParsePlatform/Parse-SDK-iOS-OSX" as a Cartfile, finally, I run carthage update]

I ran the same thing with no errors.

**Comment by fbartho** [comment]
```fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook🍣  rm -rf tmp
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook🍣  mkdir tmp; cp Cartfile.tmp tmp/Cartfile; sudo rm -rf ~/Library/Caches/org.carthage.CarthageKit/; cd tmp
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook/tmp🍣  ls
Cartfile
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook/tmp🍣  cat Cartfile 
github "ParsePlatform/Parse-SDK-iOS-OSX"
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook/tmp🍣  carthage update
*** Cloning Parse-SDK-iOS-OSX
*** Cloning Bolts-iOS
*** Checking out Bolts-iOS at "1.6.0"
*** Downloading Parse-SDK-iOS-OSX.framework binary at "1.12.0"
*** Checking out Parse-SDK-iOS-OSX at "1.12.0"
*** xcodebuild output can be found in /var/folders/vk/lwg7x4cn30ng2dw55_7536tr0000gn/T/carthage-xcodebuild.diDKBR.log
*** Building scheme "Bolts-OSX" in Bolts.xcodeproj
*** Building scheme "Bolts-tvOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-watchOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-iOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Parse-OSX" in Parse.xcworkspace
** BUILD FAILED **


The following build commands failed:
	CompileC build/Parse.build/Release/Parse-OSX.build/Objects-normal/x86_64/PFEventuallyQueue.o Parse/Internal/PFEventuallyQueue.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
/Users/fbarthelemy/Code/cookbook/cookbook/tmp/Carthage/Checkouts/Parse-SDK-iOS-OSX/Parse/Internal/PFEventuallyQueue.m:13:9: fatal error: 'Bolts/BFExecutor.h' file not found
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	CompileC build/Parse.build/Release/Parse-OSX.build/Objects-normal/x86_64/PFEventuallyQueue.o Parse/Internal/PFEventuallyQueue.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)

fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook/tmp🍣  cat /var/folders/vk/lwg7x4cn30ng2dw55_7536tr0000gn/T/carthage-xcodebuild.diDKBR.log > xcodebuildlog.txt```
Log Link: http://cl.ly/3n2E1E1p1V23

**Comment by mdiep** [comment]
Can you try again with `github "ParsePlatform/Parse-SDK-iOS-OSX" "master"`? In this most recent attempt, you used a different version.

**Comment by fbartho** [comment]
Here you go:
```fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook🍣  rm -rf tmp
fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook🍣  mkdir tmp; cp Cartfile.tmp tmp/Cartfile; sudo rm -rf ~/Library/Caches/org.carthage.CarthageKit/; cd tmp; ls -al; cat Cartfile; carthage update
Password:
total 8
drwxr-xr-x   3 fbarthelemy  staff  102 Jan 21 17:03 .
drwxr-xr-x  14 fbarthelemy  staff  476 Jan 21 17:03 ..
-rw-r--r--@  1 fbarthelemy  staff   50 Jan 21 17:03 Cartfile
github "ParsePlatform/Parse-SDK-iOS-OSX" "master"
*** Cloning Parse-SDK-iOS-OSX
*** Cloning Bolts-iOS
*** Checking out Parse-SDK-iOS-OSX at "a4eb075e6ccab0bbf645432570344e3d2afc2896"
*** Checking out Bolts-iOS at "1.6.0"
*** xcodebuild output can be found in /var/folders/vk/lwg7x4cn30ng2dw55_7536tr0000gn/T/carthage-xcodebuild.6VUPdm.log
*** Building scheme "Bolts-OSX" in Bolts.xcodeproj
*** Building scheme "Bolts-watchOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-tvOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-iOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Parse-OSX" in Parse.xcworkspace
** BUILD FAILED **


The following build commands failed:
	CompileC build/Parse.build/Release/Parse-OSX.build/Objects-normal/x86_64/PFEventuallyPin.o Parse/Internal/PFEventuallyPin.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
/Users/fbarthelemy/Code/cookbook/cookbook/tmp/Carthage/Checkouts/Parse-SDK-iOS-OSX/Parse/PFObject.h:12:9: fatal error: 'Bolts/BFTask.h' file not found
A shell task failed with exit code 65:
** BUILD FAILED **


The following build commands failed:
	CompileC build/Parse.build/Release/Parse-OSX.build/Objects-normal/x86_64/PFEventuallyPin.o Parse/Internal/PFEventuallyPin.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)

fbarthelemy@SFO-M-2849:~/Code/cookbook/cookbook/tmp🍣  cat /var/folders/vk/lwg7x4cn30ng2dw55_7536tr0000gn/T/carthage-xcodebuild.6VUPdm.log >xcodebuildlog2.txt
```
Xcode build log: http://cl.ly/0i0C231g1b3t

**Comment by mdiep** [comment]
I can't reproduce this, so I'm not sure what to tell you.

```
$ carthage update
*** Cloning Parse-SDK-iOS-OSX
*** Cloning Bolts-iOS
*** Checking out Bolts-iOS at "1.6.0"
*** Checking out Parse-SDK-iOS-OSX at "a4eb075e6ccab0bbf645432570344e3d2afc2896"
*** xcodebuild output can be found in /var/folders/xj/sydx1wkd0p147x0k_7v667940000gn/T/carthage-xcodebuild.ZM6yqS.log
*** Building scheme "Bolts-OSX" in Bolts.xcodeproj
*** Building scheme "Bolts-tvOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-watchOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Bolts-iOS-Dynamic" in Bolts.xcodeproj
*** Building scheme "Parse-watchOS-Dynamic" in Parse.xcworkspace
*** Building scheme "Parse-OSX" in Parse.xcworkspace
*** Building scheme "Parse-tvOS-Dynamic" in Parse.xcworkspace
*** Building scheme "Parse-iOS-Dynamic" in Parse.xcworkspace
$
```

**Comment by fbartho** [comment]
I'm also seeing this occur with a Cartfile that just contains:
```
github "ReSwift/ReSwiftRouter"
```
This Carthage dependency relies on "ReSwift/ReSwift" and even though building ReSwift succeeds, ReSwiftRouter can't find it unless I apply this patch:

```
diff --git a/ReSwiftRouter.xcodeproj/project.pbxproj b/ReSwiftRouter.xcodeproj/project.pbxproj
index 2816e4e..2d8c355 100644
--- a/ReSwiftRouter.xcodeproj/project.pbxproj
+++ b/ReSwiftRouter.xcodeproj/project.pbxproj
@@ -1182,6 +1182,7 @@
 				DEBUG_INFORMATION_FORMAT = dwarf;
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
 				ENABLE_TESTABILITY = YES;
+				FRAMEWORK_SEARCH_PATHS = "${PROJECT_DIR}/Carthage/Build/**";
 				GCC_C_LANGUAGE_STANDARD = gnu99;
 				GCC_DYNAMIC_NO_PIC = NO;
 				GCC_NO_COMMON_BLOCKS = YES;
@@ -1230,6 +1231,7 @@
 				DEBUG_INFORMATION_FORMAT = "dwarf-with-dsym";
 				ENABLE_NS_ASSERTIONS = NO;
 				ENABLE_STRICT_OBJC_MSGSEND = YES;
+				FRAMEWORK_SEARCH_PATHS = "${PROJECT_DIR}/Carthage/Build/**";
 				GCC_C_LANGUAGE_STANDARD = gnu99;
 				GCC_NO_COMMON_BLOCKS = YES;
 				GCC_WARN_64_TO_32_BIT_CONVERSION = YES;
```
Then Xcode builds just fine.

**Comment by fbartho** [comment]
Issue Resolved!

Additional Requirements for Carthage to find the nested dependencies:
In Xcode Settings -> Locations
1. You must have Derived Data Directory set to "Default" ![You must have Derived Data Directory set to "Default"](http://cl.ly/1V0W3o3y3l0q/download/Image%202016-02-18%20at%204.35.56%20PM.png)
2. You must have the Derived Data -> Advanced (Build Location) setting set to "Unique" ![You must have the Derived Data -> Advanced (Build Location) setting set to "Unique"](http://cl.ly/0J471m0Z1x3c/download/Image%202016-02-18%20at%204.49.57%20PM.png)

In my case, I tested with multiple values for the directory path, but I did not mess with "Advanced" and it was set to "Legacy"

This might be worth documenting? These are the same constraints that limit using custom frameworks in Playgrounds. (Discovered fixing that).
